packages/gatsby-source-contentful/src/gatsby-node.js

```diff
@@ -112,6 +119,121 @@ exports.sourceNodes = async (
     pluginConfig,
   })
 
+  createTypes(`
+  interface ContentfulEntry @nodeInterface {
+    contentful_id: String!
+    id: ID!
+  }
+`)
+
+  createTypes(`
+  interface ContentfulReference {
+    contentful_id: String!
+    id: ID!
+  }
+`)
+
+  createTypes(
+    schema.buildObjectType({
+      name: `ContentfulAsset`,
+      fields: {
+        contentful_id: { type: `String!` },
+        id: { type: `ID!` },
+      },
+      interfaces: [`ContentfulReference`, `Node`],
+    })
+  )
+
+  const gqlTypes = contentTypeItems.map(contentTypeItem =>
+    schema.buildObjectType({
+      name: _.upperFirst(
+        _.camelCase(
+          `Contentful ${
+            pluginConfig.get(`useNameForId`)
+              ? contentTypeItem.name
+              : contentTypeItem.sys.id
+          }`
+        )
+      ),
+      fields: {
+        contentful_id: { type: `String!` },
+        id: { type: `ID!` },
+      },
+      interfaces: [`ContentfulReference`, `ContentfulEntry`, `Node`],
+    })
+  )
+  createTypes(gqlTypes)
+
+  console.time(`Process Contentful data`)
+
+  // Remove deleted entries and assets from cached sync data set
+  previousSyncData.entries = previousSyncData.entries.filter(
```

Comment:
`O(n*m)` perf warning.

This is doing a `.filter` on `previousSyncData.length` elements and for every element it does an expensive `findIndex` through `currentSyncData.deletedEntries.length` elements. It _might_ be more efficient to create `Set` or `Map` caches for these to speed up the lookups.

I don't know how large these sets might generally get.

Same for `previousSyncData.assets =` below